icu4c/source/test/cintltst/cloctst.c

```diff
@@ -6665,6 +6667,155 @@ static void test_unicode_define(const char *namech, char ch, const char *nameu,
   }
 }
 
+static void checkTerminating(
+    const char* locale, const char* inLocale) {
+  UErrorCode status = U_ZERO_ERROR;
+  int32_t preflight_length = uloc_getDisplayName(
+      locale, inLocale, NULL, 0, &status);
+  if (status != U_BUFFER_OVERFLOW_ERROR) {
+      log_err("uloc_getDisplayName(%s, %s) preflight failed", locale, inLocale);
+  }
+  UChar buff[256];
+  const UChar sentinal1 = 0x6C38; // 永- a Han unicode as sentinal.
+  const UChar sentinal2 = 0x92D2; // 鋒- a Han unicode as sentinal.
+
+  // 1. Test when we set the maxResultSize to preflight_length + 1.
+  // Set sentinal1 in the buff[preflight_length-1] to check it will be replaced
+  // with display name.
+  buff[preflight_length-1] = sentinal1;
+  // Set sentinal2 in the buff[preflight_length] to check it will be replaced
+  // by null.
+  buff[preflight_length] = sentinal2;
+  // It should be properly null terminated at buff[preflight_length].
+  status = U_ZERO_ERROR;
+  int32_t length = uloc_getDisplayName(
+      locale, inLocale, buff, preflight_length + 1, &status);
+  UErrorCode convert_status = U_ZERO_ERROR;
+  // Convert the actual result to UTF8 for debugging output.
```

Comment:
UTF-8 does not work for platforms where that is not the system encoding. Please look at other test code for what function to use to convert to the system charset.

Alternatively, log the locale IDs and the output length but not the string itself.

Reply:
> Alternatively, log the locale IDs and the output length but not the string itself.

I prefer we print out the output in debugging so we can see what happen. The position of the null terminator is the key issue of this bug and seeing where the string got truncated is a very helpful information about the error.

Reply:
> UTF-8 does not work for platforms where that is not the system encoding. Please look at other test code for what function to use to convert to the system charset.

I cannot figure out which C api I can use for that. Could we just keep it in UTF8 and let the Linux version help us figure out what happen.

Reply:
I think you'd want to use something like `u_austrcpy`